pvlib/solarposition.py

```diff
@@ -1327,8 +1327,7 @@ def hour_angle(times, longitude, equation_of_time):
     equation_of_time_Spencer71
     equation_of_time_pvcdrom
     """
-    hours = np.array([(t - t.tz.localize(
-        dt.datetime(t.year, t.month, t.day)
-    )).total_seconds() / 3600. for t in times])
-    timezone = times.tz.utcoffset(times).total_seconds() / 3600.
+    tz_info = times.tz
+    timezone = tz_info.utcoffset(times).total_seconds() / 3600.
```

Comment:
is this the expected behavior?

```python
In [15]: times = pd.DatetimeIndex(start='20180101', end='20190101', freq='1s', tz='America/Phoenix')

In [16]: timezone = times.tz.utcoffset(times).total_seconds() / 3600.

In [17]: timezone
Out[17]: -7.466666666666667
```

I am surprised because 1. I expected -7 and 2. `times` is a vector so I expected to get a result for each time (offset could change for DST).

Is this preferable?

```python
In [27]: np.array((times.tz_localize(None).astype(int) - times.astype(int)) / 1e9 / 3600)
Out[27]: array([-7., -7., -7., ..., -7., -7., -7.])
```

The `np.array` wrapper ensures a known type instead of a version-dependent Index/array type.

Reply:
FYI: it's only used 2 places:
```
Searching 160 files for "utcoffset"

C:\Users\mikm\Projects\pvlib-python\pvlib\clearsky.py:
  664          * automatically determines sample_interval
  665          * requires a reference clear sky series instead calculating one
  666:           from a user supplied location and UTCoffset
  667          * parameters are controllable via keyword arguments
  668          * option to return individual test components and clearsky scaling

C:\Users\mikm\Projects\pvlib-python\pvlib\solarposition.py:
  547          # pyephem drops timezone when converting to its internal datetime
  548          # format, so handle timezone explicitly here
  549:         obs.date = ephem.Date(thetime - thetime.utcoffset())
  550          sunrise.append(_ephem_to_timezone(rising(sun), time.tz))
  551          sunset.append(_ephem_to_timezone(setting(sun), time.tz))
  ...
 1329      """
 1330      tz_info = times.tz
 1331:     timezone = tz_info.utcoffset(times).total_seconds() / 3600.
 1332      hours = 1 / (3600. * 1.e9) * (
 1333          times.astype(np.int64) - times.normalize().astype(np.int64))
```

So [`utcoffset()` is also a Python builtin](https://docs.python.org/3/library/datetime.html#datetime.datetime.utcoffset) which is overloaded by PyTz. The difference is that if `utcoffset` is called on a Python datetime then it will return the utcoffset of it's tz info called on itself. So in our example above we could have called:

    >>> dt.utcoffset()
    datetime.timedelta(-1, 59520)

Also FYI: there is an [inefficient loop in `rise_set_transit_ephem`](https://github.com/pvlib/pvlib-python/blob/master/pvlib/solarposition.py#L545-L552) which uses slow `to_pydatetime()` and `utcoffset()`, but it's probably addressed in #588?

Other than that I think there's only 2 other instances of `utcoffset` this one in hour_angle and one in #583. I will fix both. Thanks for catching this!

Reply:
FYI: this is also an example of why we should *never* replace tzinfo manually, but **always** use the `localize()` method so that the correct timezone from the transition info table is used.

Reply:
Thanks for looking into it. I've seen this kind of funny behavior with the AZ timezones before but it didn't occur to me that it could be a problem in this function. Some other common timezones have similar behavior (though I don't remember which).

Reply:
To be clear, I don't think it's a problem with the function `utcoffset()`. Honestly, I think this was my mistake for using `utcoffset()` with a sequence of times like `pandas.DatetimeIndex`, because there could potentially be a different timezone for each item in the sequence, so it's ambiguous which timezone should be used. Therefore, I think the problem is that the `tz` (aka `tzinfo`) that's set in the `DatetimeIndex` is unreliable and ambiguous, or the problem could be with `utcoffset` which should be overloaded to return a list of timezones localized to each item in the sequence.